### PR TITLE
libtest: run all tests in their own thread, if supported by the host

### LINKED ISSUE #59122: Rust 1.33.0 failing UI tests when run with single core
I'm trying to package/build Rust 1.33.0 on Guix and I'm consistently getting some test failures regarding panic messages.

Several run-fail tests check that a `#[test]` with a name like `test_foo` prints a message like `thread 'test_foo' panicked...`, but instead the output will come out as `thread 'main' panicked...`.

I'm not really sure what can be the cause of the issue, but FWIW we build Rust with the previous stable version already packaged in Guix, and we'll link to our packaged versions of llvm and jemalloc.

Happy to provide more details as needed; log from the failing test cases is below

<details>
<pre>
  < Compiletest { compiler: Compiler { stage: 2, host: "x86_64-unknown-linux-gnu" }, target: "x86_64-unknown-linux-gnu", mode: "compile-fail", suite: "compile-fail", path: Some("src/test/compile-fail"), compare_mode: None }
< CompileFail { compiler: Compiler { stage: 2, host: "x86_64-unknown-linux-gnu" }, target: "x86_64-unknown-linux-gnu" }
c Assemble { target_compiler: Compiler { stage: 2, host: "x86_64-unknown-linux-gnu" } }
> RunFail { compiler: Compiler { stage: 2, host: "x86_64-unknown-linux-gnu" }, target: "x86_64-unknown-linux-gnu" }
  > Compiletest { compiler: Compiler { stage: 2, host: "x86_64-unknown-linux-gnu" }, target: "x86_64-unknown-linux-gnu", mode: "run-fail", suite: "run-fail", path: Some("src/test/run-fail"), compare_mode: None }
    c Test { target: "x86_64-unknown-linux-gnu", compiler: Compiler { stage: 2, host: "x86_64-unknown-linux-gnu" } }
    c Test { target: "x86_64-unknown-linux-gnu", compiler: Compiler { stage: 2, host: "x86_64-unknown-linux-gnu" } }
    c TestHelpers { target: "x86_64-unknown-linux-gnu" }
    c TestHelpers { target: "x86_64-unknown-linux-gnu" }
    c RemoteCopyLibs { compiler: Compiler { stage: 2, host: "x86_64-unknown-linux-gnu" }, target: "x86_64-unknown-linux-gnu" }
    c Assemble { target_compiler: Compiler { stage: 0, host: "x86_64-unknown-linux-gnu" } }
    c Compiletest { compiler: Compiler { stage: 0, host: "x86_64-unknown-linux-gnu" }, target: "x86_64-unknown-linux-gnu" }
    c Assemble { target_compiler: Compiler { stage: 0, host: "x86_64-unknown-linux-gnu" } }
    c Llvm { target: "x86_64-unknown-linux-gnu", emscripten: false }
    c Sysroot { compiler: Compiler { stage: 2, host: "x86_64-unknown-linux-gnu" } }
    c Libdir { compiler: Compiler { stage: 2, host: "x86_64-unknown-linux-gnu" }, target: "x86_64-unknown-linux-gnu" }
    c Sysroot { compiler: Compiler { stage: 2, host: "x86_64-unknown-linux-gnu" } }
    c Llvm { target: "x86_64-unknown-linux-gnu", emscripten: false }
Check compiletest suite=run-fail mode=run-fail (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
running: "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/src/test/run-fail" "--build-base" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/test/run-fail" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "run-fail" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/gnu/store/bvnj04sqiwc1gcis29swk6likyc5dw8f-llvm-6.0.1/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/gnu/store/g87hamjyipk1j6dfq5pjfzfnfb64spbv-python2-2.7.15/bin/python2" "--lldb-python" "/gnu/store/g87hamjyipk1j6dfq5pjfzfnfb64spbv-python2-2.7.15/bin/python2" "--gdb" "/gnu/store/pp6nvrw47ip2r54k4kyk5a8fhyxvhsb8-gdb-8.2/bin/gdb" "--verbose" "--quiet" "--llvm-version" "6.0.1\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" ""
configuration:
compile_lib_path: "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/stage2/lib"
run_lib_path: "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib"
rustc_path: "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/stage2/bin/rustc"
rustdoc_path: None
src_base: "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/src/test/run-fail"
build_base: "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/test/run-fail"
stage_id: stage2-x86_64-unknown-linux-gnu
mode: run-fail
run_ignored: false
filter: (none)
filter_exact: false
runtool: (none)
host-rustcflags: -Crpath -O -Zunstable-options  -Lnative=/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/native/rust-test-helpers
target-rustcflags: -Crpath -O -Zunstable-options  -Lnative=/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/native/rust-test-helpers
target: x86_64-unknown-linux-gnu
host: x86_64-unknown-linux-gnu
android-cross-path: ""
adb_path: "adb"
adb_test_dir: "/data/tmp/work"
adb_device_status: false
ar: ar
linker: None
verbose: true
quiet: true

running 143 tests
.................................................................................................... 100/143
........................FF.................
failures:

---- [run-fail] run-fail/test-panic.rs stdout ----

executing "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/src/test/run-fail/test-panic.rs" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/test/run-fail/test-panic/a" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--test" "-L" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/test/run-fail/test-panic/auxiliary"
------stdout------------------------------

------stderr------------------------------

------------------------------------------
executing "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/test/run-fail/test-panic/a"
------stdout------------------------------

running 1 test
test test_foo ... FAILED

failures:

---- test_foo stdout ----
thread 'main' panicked at 'explicit panic', /tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/src/test/run-fail/test-panic.rs:8:5
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

failures:
    test_foo

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

------stderr------------------------------

------------------------------------------

error: error pattern 'thread 'test_foo' panicked at' not found!
status: exit code: 101
command: "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/test/run-fail/test-panic/a"
stdout:
------------------------------------------

running 1 test
test test_foo ... FAILED

failures:

---- test_foo stdout ----
thread 'main' panicked at 'explicit panic', /tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/src/test/run-fail/test-panic.rs:8:5
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

failures:
    test_foo

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

------------------------------------------
stderr:
------------------------------------------

------------------------------------------

thread '[run-fail] run-fail/test-panic.rs' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:3245:9

---- [run-fail] run-fail/test-should-fail-bad-message.rs stdout ----

executing "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/src/test/run-fail/test-should-fail-bad-message.rs" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/test/run-fail/test-should-fail-bad-message/a" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--test" "-L" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/test/run-fail/test-should-fail-bad-message/auxiliary"
------stdout------------------------------

------stderr------------------------------

------------------------------------------
executing "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/test/run-fail/test-should-fail-bad-message/a"
------stdout------------------------------

running 1 test
test test_foo ... FAILED

failures:

---- test_foo stdout ----
thread 'main' panicked at 'blah', /tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/src/test/run-fail/test-should-fail-bad-message.rs:9:5
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
note: Panic did not include expected string 'foobar'

failures:
    test_foo

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

------stderr------------------------------

------------------------------------------

error: error pattern 'thread 'test_foo' panicked at' not found!
status: exit code: 101
command: "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/test/run-fail/test-should-fail-bad-message/a"
stdout:
------------------------------------------

running 1 test
test test_foo ... FAILED

failures:

---- test_foo stdout ----
thread 'main' panicked at 'blah', /tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/src/test/run-fail/test-should-fail-bad-message.rs:9:5
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
note: Panic did not include expected string 'foobar'

failures:
    test_foo

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

------------------------------------------
stderr:
------------------------------------------

------------------------------------------

thread '[run-fail] run-fail/test-should-fail-bad-message.rs' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:3245:9

failures:
    [run-fail] run-fail/test-panic.rs
    [run-fail] run-fail/test-should-fail-bad-message.rs

test result: FAILED. 141 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out

thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:495:22

command did not execute successfully: "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/src/test/run-fail" "--build-base" "/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/test/run-fail" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "run-fail" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/gnu/store/bvnj04sqiwc1gcis29swk6likyc5dw8f-llvm-6.0.1/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/tmp/guix-build-rust-1.33.0.drv-0/rustc-1.33.0-src/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/gnu/store/g87hamjyipk1j6dfq5pjfzfnfb64spbv-python2-2.7.15/bin/python2" "--lldb-python" "/gnu/store/g87hamjyipk1j6dfq5pjfzfnfb64spbv-python2-2.7.15/bin/python2" "--gdb" "/gnu/store/pp6nvrw47ip2r54k4kyk5a8fhyxvhsb8-gdb-8.2/bin/gdb" "--verbose" "--quiet" "--llvm-version" "6.0.1\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" ""
expected success, got: exit code: 101
</pre></details>

### LINKED ISSUE #70492: Tests do not set thread name when concurrency is disabled
Currently there are only workarounds for getting the test name from the current running test. The main one that people use is the thread name, but that does not work when tests are run with `--test-threads=1`.

I believe it would be beneficial to spawn a thread anyways so that the test name is discoverable. Alternatively — and maybe that's the better solution — there should be an API to discover the name of the current test.

### PR DESCRIPTION
**Pull request by RalfJung**
This reverts the threading changes of https://github.com/rust-lang/rust/pull/56243, which made it so that with `-j1`, the test harness does not spawn any threads. Those changes were done to enable Miri to run the test harness, but Miri supports threads nowadays, so this is no longer needed. Using a thread for each test is useful because the thread's name can be set to the test's name which makes panic messages consistent between `-j1` and `-j2` runs and also a bit more readable.

I did not revert the HashMap changes of https://github.com/rust-lang/rust/pull/56243; using a deterministic map seems fine for the test harness and the more deterministic testing is the better.

Fixes https://github.com/rust-lang/rust/issues/59122
Fixes https://github.com/rust-lang/rust/issues/70492

### CODE REVIEW

**Comment by RalfJung** [comment]
I'll need to do some testing to ensure the test harness still works on Windows, which has less complete threading support in Miri than Unix targets.

**Comment by RalfJung** [comment]
Unfortunately on the Miri side this is blocked on https://github.com/rust-lang/miri/issues/2628.

**Comment by the8472** [comment]
> Using a thread for each test is useful because the thread's name can be set to the test's name

On many-core systems it could be useful to use a thread pool rather than a thread-per-test to avoid the thread-spawning costs (e.g. contention on the memory-map lock for the thread stack creation). Maybe updating the thread name between tests could work?

**Comment by RalfJung** [comment]
> On many-core systems it could be useful to use a thread pool rather than a thread-per-test to avoid the thread-spawning costs (e.g. contention on the memory-map lock for the thread stack creation).

Sure, but that's a new feature I won't implement. All this here does is revert a PR I wrote years ago. It doesn't even change behavior for many-core systems.

> Maybe updating the thread name between tests could work?

The standard library doesn't have an API for that.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/miri/pull/2630 adds the missing thread parker support for Windows. :)

**Comment by RalfJung** [comment]
All right, with https://github.com/rust-lang/rust/pull/103721 on its way this is ready for review (but shouldn't land before that Miri update lands).
r? @thomcc

**Review by thomcc — APPROVED** [review]
Huh, I had no idea the reason miri was the reason for this part of libtest's cruftiness.

Really glad to see it gone either way, it vaguely was in the way of a few other things I've wanted to do for a while now.

r=me once you're ready to land it.

**Comment by RalfJung** [comment]
Thanks! https://github.com/rust-lang/rust/pull/103721 has landed, so this is good to go.

@bors r=thomcc

**Comment by bjorn3** [comment]
What about inherently single threaded environments like wasm? Won't this break libtest for those? As far as I am aware wasm32-wasi currently supports using libtest.

**Comment by bjorn3** [comment]
Missed that, my bad.

**Comment by knopp** [comment]
Rust default test harness doesn't really work for anything on macOS that requires execution on main thread. Unfortunately main thread is special (so that on some places there are actually checks for pthread_main_np, i.e. in libdispatch. It would be immensely helpful if there was way to run tests with the default harness directly on main thread.

**Comment by RalfJung** [comment]
@knopp comments on old merged MRs are unlikely to lead anywhere. Please open a new issue to describe the problem.

**Comment by RalfJung** [comment]
See e.g. https://github.com/rust-lang/rust/issues/104053

**Pull request merged**
